### ISSUE TITLE
example of language-level support for interpreter contracts

### ISSUE THREAD

**Issue by nmushegian** [nexusdev/dapple-script#2]
- How does it look to specify running a function `via` an interpreter?
- How does it look to mark a function as "atomic only" ?

**Comment by mhhf** [comment]
Syntax wise I like to stay as close as possible to solidity to reduce the cognitive distance. Atomic only functions could be marked with an _modifier_ equivalent (e.g. `atomic`). Calling non atomic functions in an atomic fashion could be done with a predicate: `offChain.atomic()`:

```
...
  function onChain() atomic { ... }
  function offChain() { ... }
  function call() {
    // execute the offChain() sequence off chain:
    offChain()
    // execute the offChain() sequence on chain:
    offChain.atomic()
  }
...
```

**Comment by nmushegian** [comment]
Which interpreter address is it using? Maybe `doStuff(args).atomic(via_which_address);` ?

**Comment by mhhf** [comment]
Ah, this is a good point! In my model an interpreter would be either an library with a fixed address defined by default in .dapplerc or dappfile or an transit contract which gets deployed, executed and suicided in one run.
Why would an user want do point against custom interpreter?

EDIT: It suicides only if the run can be finished, otherwise the state will get saved it awaits further instructions.

**Comment by nmushegian** [comment]
> an library with a fixed address defined by default

If you call a library from a key it's just a contract call, this would mean this universal singleton would have to manage "sessions" of different users.

> an transit contract which gets deployed, executed and suicided in one run.

This option seems best. One key can manage multiple totally independent interpreters. The transit contract could be instantiated with state from the off-chain scope. Neat!

**Comment by nmushegian** [comment]
The reason to re-use interpreters is to maintain the right `msg.sender` across multiple transactions (e.g. two steps of atomic update operations). It still seems there should be a way to distinguish between interpreters, then.

**Comment by mhhf** [comment]
> If you call a library from a key it's just a contract call, this would mean this universal singleton would have to manage "sessions" of different users.

Yo are right! We can go for transit contracts for state management then. 

> `doStuff(args).atomic(via_which_address);`

I somehow dislike the idea of _scripts_ being state dependent. 

In case we go with the transit contracts: the sate for off chain executions as well information about on chain interpreters should be stored in the local environment, e.g. `./.dapple/UTC-10-04-2016-13-23.deploy.step` and managed via dapples cli (e.g. listing, resuming or running new scripts against saved interpreter states via `dapple step`)

(can you provide an example, where it would be necessary to pause and resuming an atomic interpreter for something different then out-of-gas reasons? Maybe with some complex intercommunication...)

To abstract `msg.sender` away or if there are scenarios where a script has to interact with multiple interpreters we can use **actors** much like we would use an authorized `Tester` contract. I'd suggest opening another design issue for this. Here we also can come up with a minimal implementation of the actor model for intercommunication and persistent state.  

I need to spend more time to think about a clean way to represent actors, but would favor separating this concept from scripts.

**Comment by nmushegian** [comment]
> the sate for off chain executions as well information about on chain interpreters should be stored in the local environment, e.g. ./.dapple/UTC-10-04-2016-13-23.deploy.step and managed via dapples cli (e.g. listing, resuming or running new scripts against saved interpreter states via `dapple step`

yes! This has always matched my intuition about where this would end up

> can you provide an example, where it would be necessary to pause and resuming an atomic interpreter for something different then out-of-gas reasons? Maybe with some complex intercommunication...

block gas limit is different from out-of-gas. There might be some O(n) operations you have to split across blocks as best as you can, for example. As for complex intercom: what about async/sharding? Continuation definitions have to be on-chain

**Comment by nmushegian** [comment]
Separating actors from the language level is fine, I just really want to capture the duality between code run by actors and "code" run off-chain. There will be on-chain async primitives for sure and at that point there's almost no difference except for who can act as tx.origin

**Comment by mhhf** [comment]
If your fine with this I'd stick with simple step, pause, resume on-chain state and with simple single actor managed by dapple step cli interface for now (`dapple run deploy.ds --actor "0xff..ff`). I see how dapple would benefit from async/ sharding and a complete actor system. We can continue designing this after finishing the other tasks like simple interpreter, chain forking and dapphub.

**Comment by nmushegian** [comment]
`dapple run deploy.ds --actor "0xff..ff`

This seems fine to me.
